lib/usdUfe/ufe/UsdClipboardHandler.cpp

```diff
@@ -0,0 +1,212 @@
+
+UsdClipboardHandler::UsdClipboardHandler() { _clipboard = std::make_shared<UsdClipboard>(); }
+
+/*static*/
+UsdClipboardHandler::Ptr UsdClipboardHandler::create()
+{
+    return std::make_shared<UsdClipboardHandler>();
+}
+
+//------------------------------------------------------------------------------
+// Ufe::ClipboardHandler overrides
+//------------------------------------------------------------------------------
+
+Ufe::UndoableCommand::Ptr UsdClipboardHandler::cutCmd_(const Ufe::Selection& selection)
+{
+    return UsdCutClipboardCommand::create(selection, _clipboard);
+}
+
+Ufe::UndoableCommand::Ptr UsdClipboardHandler::copyCmd_(const Ufe::Selection& selection)
+{
+    return UsdCopyClipboardCommand::create(selection, _clipboard);
+}
+
+Ufe::PasteClipboardCommand::Ptr
+UsdClipboardHandler::pasteCmd_(const Ufe::SceneItem::Ptr& parentItem)
+{
+    return UsdPasteClipboardCommand::create(parentItem, _clipboard);
+}
+
+Ufe::UndoableCommand::Ptr UsdClipboardHandler::pasteCmd_(const Ufe::Selection& parentItems)
+{
+    return UsdPasteClipboardCommand::create(parentItems, _clipboard);
+}
+
+bool UsdClipboardHandler::hasItemsToPaste_()
+{
+    if (auto clipboardStage = _clipboard->getClipboardData()) {
+        for (auto prim : clipboardStage->Traverse()) {
+            // Return true if there is at least one valid prim to paste.
+            if (prim)
+                return true;
+        }
+    }
+    return false;
+}
+
+bool UsdClipboardHandler::canBeCut_(const Ufe::SceneItem::Ptr& item)
+{
+    auto usdItem = std::dynamic_pointer_cast<UsdSceneItem>(item);
+    if (!usdItem)
+        return false;
+
+    const auto prim = usdItem->prim();
+    if (!prim)
+        return false;
+
+    const auto primParent = prim.GetParent();
+    if (!primParent)
+        return false;
+
+    // Special cut conditions that only applies to UsdShadeNodeGraph and UsdShadeShader.
+    if (PXR_NS::UsdShadeNodeGraph(prim) || PXR_NS::UsdShadeShader(prim)) {
+        const auto primAttrs = prim.GetAuthoredAttributes();
+        for (const auto& attr : primAttrs) {
+            const auto kBaseNameAndType
+                = PXR_NS::UsdShadeUtils::GetBaseNameAndType(PXR_NS::TfToken(attr.GetName()));
+
+            if (kBaseNameAndType.second != PXR_NS::UsdShadeAttributeType::Output
+                && kBaseNameAndType.second != PXR_NS::UsdShadeAttributeType::Input) {
+                continue;
+            }
+
+            if (kBaseNameAndType.second == PXR_NS::UsdShadeAttributeType::Input) {
+                // The attribute could be a destination for connected sources, so check for its
+                // connections.
+                PXR_NS::UsdShadeSourceInfoVector sourcesInfo
+                    = pxr::UsdShadeConnectableAPI::GetConnectedSources(attr);
+
+                if (!sourcesInfo.empty()) {
+                    return false;
+                }
+            }
+
+            if (kBaseNameAndType.second == PXR_NS::UsdShadeAttributeType::Output) {
```

Comment:
IDK why output are only checked for siblings... must be a LookdevX thing.

Reply:
@alicedegirolamo ??

Reply:
I added this as a code comment.